src/Markup/Avalonia.Markup.Xaml.Loader/CompilerExtensions/Visitors/NameScopeRegistrationVisitor.cs

```diff
@@ -0,0 +1,47 @@
+using System.Collections.Generic;
+using Avalonia.Markup.Xaml.XamlIl.CompilerExtensions.Transformers;
+using XamlX.Ast;
+using XamlX.TypeSystem;
+
+namespace Avalonia.Markup.Xaml.XamlIl.CompilerExtensions.Visitors;
+
+internal class NameScopeRegistrationVisitor : Dictionary<string, (IXamlType type, IXamlLineInfo line)>, IXamlAstVisitor
```

Comment:
It's not quite clear what this class is supposed to do.

It appears to count `AvaloniaNameScopeRegistrationXamlIlNode` but only if they are in the first inner name scope? 

When exactly does that scope gets introduced for a resource?

I'd suggest to at least change `initialMetadataScopeLevel = 0` to `targetMetadataScopeLevel` which would be 2 in this scenario.

Reply:
The same visitor I have used for TemplatePart PR. So, it's definitely going to be reused at least in two places. 

> It appears to count AvaloniaNameScopeRegistrationXamlIlNode but only if they are in the first inner name scope?

Not just count, but all of the x:Name registration names and types. I needed type info for TemplatePart PR, but can be kept here as well. See https://github.com/AvaloniaUI/Avalonia/pull/14180
Only first name scope level, because any nested namescope is isolated from the parent, and cannot be accessed from the same context. I assume in another situation it could be required, but not now.

> I'd suggest to at least change initialMetadataScopeLevel = 0 to targetMetadataScopeLevel which would be 2 in this scenario.

Yeah, can do that.

Reply:
Ah, that's why the _initial_ scope level is 1. It's the class API being confusing.

Reply:
Yeah, NameScopeRegistrationVisitor itself has initial scope of "0", which is expected.
But resources transformer doesn't start from the root, so it assumes we are in the middle of the tree. I will add a comment to clarify that.